### PR TITLE
Enable Deepseek-v4 hybrid_model in dev branch Part (1/N)

### PR DESCRIPTION
**Pull request by guihong-nv**
# What does this PR do ?
Enable Deepseek-v4 hybrid_model in dev branch

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by guihong-nv** [comment]
/ok to test [2d28355](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/2d28355ae0d6291a25528dfde5895995c50300d2)

**Comment by guihong-nv** [comment]
/ok to test [c179e58](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/c179e58947d2f340e922e758aaf305b3c4dc5760)

**Comment by guihong-nv** [comment]
/ok to test [45e2efc](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/45e2efc81e6a13d56f2e5566c53035684b65138a)

**Review by Phlip79 — APPROVED** [review]

**Comment by guihong-nv** [comment]
/ok to test [59f8174](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/59f817415556f6770587818f9b3769d779befc11)

File: megatron/core/transformer/hyper_connection.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
     from megatron.core.tensor_parallel.random import CheckpointManager
 
 
-@torch.compile
+# @torch.compile  # disabled 2026-05-27 — Dynamo shape mismatch on hybrid_dev mHC path
```
Comment (hxbai) [inline megatron/core/transformer/hyper_connection.py]: Why are they disabled?

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -1433,9 +1433,14 @@ def __post_init__(self):
...
             mtp_layers = self.mtp_num_layers or 0
+            # Minimum length is num_layers + mtp_num_layers (the GPT path uses exactly this,
+            # where mtp_num_layers == #MTP transformer layers). On HybridModel an MTP "depth"
+            # can expand to MULTIPLE hybrid layers, so mtp_num_layers (= depth count) undercounts
+            # the real MTP layers and csa_compress_ratios must be at least long enough to index
+            # every MTP attention layer (num_layers + layer_number - 1). Hence ">=", not "==".
             expected_len = self.num_layers + mtp_layers
-            assert len(self.csa_compress_ratios) == expected_len, (
-                f"csa_compress_ratios length ({len(self.csa_compress_ratios)}) must equal "
+            assert len(self.csa_compress_ratios) >= expected_len, (
```
Comment (hxbai) [inline megatron/core/transformer/transformer_config.py:1478]: This is error-prune. Could you calculate the exact needed length and use `==`?

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -2097,8 +2097,37 @@ def core_transformer_config_from_args(args, config_class=None):
...
+            # One ratio entry per ACTUAL layer: main layers, then every MTP layer of every MTP
+            # depth (a depth can contain multiple hybrid layers, e.g. "/MD-E"). This makes the
+            # array long enough for the deepseek attn index (num_layers + layer_number - 1) for
+            # any MTP attention position, not just depth-first. C->4, H->128, others->0.
+            _sections = _pat.split(Symbols.MTP_SEPARATOR)
+            _ratios = [_ratio_map.get(c, 0) for c in _sections[0].replace(Symbols.PIPE, '')]
+            for _mtp_sec in _sections[1:]:
+                _ratios += [_ratio_map.get(c, 0) for c in _mtp_sec.replace(Symbols.PIPE, '')]
+            kw_args['csa_compress_ratios'] = _ratios
+            args.csa_compress_ratios = _ratios
```
Comment (hxbai) [inline megatron/training/arguments.py:2194]: How to set the arguments for DSv4 with HybridModel? What about ratio==0 layers?

File: megatron/core/extensions/transformer_engine.py [hunk 4]
```diff
@@ -1757,12 +1757,21 @@ def __init__(
...
+
+                _gl_params = _inspect.signature(te.pytorch.GroupedLinear.__init__).parameters
+                if "single_grouped_weight" in _gl_params:
+                    extra_kwargs["single_grouped_weight"] = getattr(
+                        config, "moe_single_grouped_weight", False
+                    )
+                if "single_grouped_bias" in _gl_params:
+                    extra_kwargs["single_grouped_bias"] = getattr(
+                        config, "moe_single_grouped_bias", False
+                    )
```
[the remark this answers is not in the document]
Comment (guihong-nv) [reply megatron/core/extensions/transformer_engine.py:2035]: FIXED: cached at module scope (introspect once); dropped the _inspect alias

File: megatron/core/models/hybrid/hybrid_block.py [hunk 5]
```diff
@@ -611,6 +662,8 @@ def get_inner_quant_context(config, layer_number):
             inp=hidden_states, requires_grad=hidden_states.requires_grad, keep_graph=True
         )
 
+        if mhc_multistream is not None:
+            return hidden_states, mhc_multistream
```
[the remark this answers is not in the document]
Comment (guihong-nv) [reply megatron/core/models/hybrid/hybrid_block.py:978]: FIXED: added `-> Union[Tensor, Tuple[Tensor, Tensor]]` annotation and a precise docstring describing the single condition under which the 2-tuple is returned; noted PP send/recv only transfers the contracted hidden_states.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 6]
```diff
@@ -73,8 +73,19 @@ def save_loss_to_tracker(
...
+        needed = max(num_layers, layer_number)
         if "values" not in tracker:
-            tracker["values"] = torch.zeros(num_layers, device=torch.cuda.current_device())
+            tracker["values"] = torch.zeros(needed, device=torch.cuda.current_device())
+        elif tracker["values"].shape[0] < needed:
+            grown = torch.zeros(
+                needed, device=tracker["values"].device, dtype=tracker["values"].dtype
+            )
+            grown[: tracker["values"].shape[0]] = tracker["values"]
+            tracker["values"] = grown
```
[the remark this answers is not in the document]
Comment (guihong-nv) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:88]: FIXED: `reduce_loss_in_tracker` reduce-MAXes the tracker size across the PP group and pads to it before the value all_reduce → identical shapes on every rank. Verified PP=2 + MTP.

[the remark this answers is not in the document]
Comment (guihong-nv) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:88]: FIXED: reduce_loss_in_tracker reduce-MAXes the tracker size across the PP group and pads to it before the value all_reduce → identical shapes on every rank. Verified PP=2 + MTP.

File: megatron/core/transformer/transformer_config.py (hunk 2 above)
Comment (guihong-nv) [reply megatron/core/transformer/transformer_config.py:1478]: Pls check the comments in line 1436 to 1440 above.

Comment (guihong-nv) [reply megatron/core/transformer/transformer_config.py:1478]: We also have a guard here https://github.com/NVIDIA/Megatron-LM/pull/5042/changes#diff-42fa19ec8893eabf951a5bb21edfc0dfe7c9c8949d5087ab133f5897ea0e3213R2126

File: megatron/training/arguments.py (hunk 3 above)
Comment (guihong-nv) [reply megatron/training/arguments.py:2194]: DSv4 hybrid is configured by the layer pattern + MLA args; csa_compress_ratios is optional: --hybrid-layer-pattern "M-MCM-MHM-D-" # M=mamba, C=CSA, H=HCA, D=DSA, -=MLP, E=MoE, |=PP, /=MTP

File: hybrid_builders.py [hunk 7]
```diff
@@ -13,6 +13,47 @@ def hybrid_builder(args, pre_process, post_process, vp_stage=None, config=None,
...
+    # v_head_dim and qk_pos_emb_head_dim (MLATransformerConfig.__post_init__ does this for the
+    # GPT path). The hybrid config can reach here without that derivation applied, which breaks
+    # the MLA up-proj / fused-rope contract (q head dim must equal qk_head_dim + qk_pos_emb_head
+    # _dim == v_head_dim). Apply it for any DSv4 MLA attention: experimental_attention_variant
+    # == dsv4_hybrid, OR the layer pattern uses a DSv4 attention symbol (D/C/H). Idempotent.
+    _pattern = getattr(args, "hybrid_layer_pattern", None) or ""
+    _uses_dsv4_attn = (
+        getattr(args, "experimental_attention_variant", None) == "dsv4_hybrid"
+        or any(sym in _pattern for sym in ("D", "C", "H"))
+    )
```
[the remark this answers is not in the document]
Comment (guihong-nv) [reply hybrid_builders.py:31]: FIXED: pattern branch now checks only DSv4 symbols ("C","H")

File: megatron/training/arguments.py (hunk 3 above)
Comment (guihong-nv) [reply megatron/training/arguments.py:2194]: I address this issue in [7685f39](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/7685f39ad55db61b9f5853f40b7b2e0a80f80e14). Pls take a look. Thanks!

**Comment by guihong-nv** [comment]
/ok to test [42681a9](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/42681a9816d7f17d4e447977c625ea91db69f624)

**Comment by guihong-nv** [comment]
/ok to test [96fd054](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/96fd0546e7cf67d34b368f90a14ae8f11afe5079)

**Review by hxbai — APPROVED** [review]

**Comment by guihong-nv** [comment]
/ok to test [d4a265f](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/d4a265f7c18a90f329faa6efa3ac457a674c1d89)

**Review by Phlip79 — APPROVED** [review]

**Comment by Connor-XY** [comment]
/ok to test [8b0f8ed](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/8b0f8ede466c93d1b095c85c29460e396f2063fd)

**Comment by guihong-nv** [comment]
/ok to test [f360b25](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/f360b255c378649bac97bbfc0d8fb5ea1c8a8f85)

**Comment by guihong-nv** [comment]
/ok to test [69abeed](https://github.com/NVIDIA/Megatron-LM/pull/5042/commits/69abeedb6e4f42396a011c21d416a87e84777577)
